### ISSUE TITLE
Improve API usability regarding xxxByYyy() methods?

### ISSUE THREAD

**Issue by AnyhowStep** [AnyhowStep/tsql#76]
I have **a lot** of methods named `xxxByYyy()`.

+ `fetchOneByCandidateKey()`
+ `fetchOneByPrimaryKey()`
+ `fetchOneBySuperKey()`
+ `existsOneByCandidateKey()`
+ `existsOneByPrimaryKey()`
+ `existsOneBySuperKey()`
+ `assertExistsOneByCandidateKey()`
+ `assertExistsOneByPrimaryKey()`
+ `assertExistsOneBySuperKey()`
+ `deleteOneByCandidateKey()`
+ `deleteOneByPrimaryKey()`
+ `deleteOneBySuperKey()`
+ `updateAndFetchOneByCandidateKey()`
+ `updateAndFetchOneByPrimaryKey()`
+ `updateAndFetchOneBySuperKey()`
+ etc.
+ etc.

The list is really long.

-----

`table`s usually implement versions of the above that are just `xxx()` and take an arbitrary `WHERE` clause.

The `ByYyy()` variants are supposed to be convenience methods for "common" use cases. For online transaction processing, it's usually doing some operation with primary keys, candidate keys, and super keys as identifiers.

-----

I was wondering if this API might be better,
```ts
myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .fetchOne(connection);

myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .exists(connection);

myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .assertExists(connection);

myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .deleteOne(connection);

myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .updateAndFetchOne(
    connection,
    columns => {
      return {
        columnToUpdate0 : concat(columns.columnToUpdate0, "-hello"),
        columnToUpdate1,
        columnToUpdate2,
      };
    }
  );
```

vs,
```ts
myTable.fetchOneByCandidateKey(
  connection,
  {
    myCandidateKey0,
    myCandidateKey1,
  }
);

myTable.existsByCandidateKey(
  connection,
  {
    myCandidateKey0,
    myCandidateKey1,
  }
);

myTable.assertExistsByCandidateKey(
  connection,
  {
    myCandidateKey0,
    myCandidateKey1,
  }
);

myTable.deleteOneByCandidateKey(
  connection,
  {
    myCandidateKey0,
    myCandidateKey1,
  }
);

myTable.updateAndFetchOneByCandidateKey(
  connection,
  {
    myCandidateKey0,
    myCandidateKey1,
  },
  columns => {
    return {
      columnToUpdate0 : concat(columns.columnToUpdate0, "-hello"),
      columnToUpdate1,
      columnToUpdate2,
    };
  }
);
```

**Comment by AnyhowStep** [comment]
Now that I look at it, I guess I'm not into the `usingCandidateKey()/usingPrimaryKey()/usingSuperKey()` idea.

**Comment by AnyhowStep** [comment]
Maybe have a different API entirely?

@G-Rath any thoughts?

**Comment by G-Rath** [comment]
@AnyhowStep I actually think your proposed API is pretty nice - it also gives you the ability to pass the result around, which could be useful for composing.

I think your pros vs cons writeup makes it pretty clear cut: while you don't have the verbs first, I think in a database/"usingX" context that's actually still somewhat a pro, b/c the verb is still in the right place for the context: You don't actually care what you're `updateAndFetchOne`ing on aside from "the data" (whatever "the data" might be in the current context).

Effectively, I think your proposed API allows for nicer encapsulation of responsibilities, and still sounds very nice: "myTable, using candidate key, fetch and update one connection".

**Comment by AnyhowStep** [comment]
I'm wondering how the, hopefully less used, API will be like if I take the `usingX` approach.

Right now, it is,
```ts
table.fetchOne(
  connection,
  //A callback function so one can craft arbitrary WHERE clauses
  columns => and(
    gtEq(columns.createdAt, new Date("2019-12-20T00:00:00Z")),
    not(columns.banned)
  )
);
```

I think this is fine, but I don't think users will be using this API in general.

Keeping it like this means copy-pasting code, though.
(Certainly wayyyyyy less copy-pasting, though).

-----

I could also try and shoehorn the proposed API here,
```ts
table
  //A callback function so one can craft arbitrary WHERE clauses
  .where(columns => and(
    gtEq(columns.createdAt, new Date("2019-12-20T00:00:00Z")),
    not(columns.banned)
  ))
  .fetchOne(connection);
```

**Comment by pachoclo** [comment]
I definitely like the...
`myTable
  .usingCandidateKey({
    myCandidateKey0,
    myCandidateKey1,
  })
  .fetchOne(connection);`
... approach much better. It looks cleaner and more pleasant to read, in my opinion. 

Just two (most likely lazy) questions:
- How difficult would it be to make `usingCandidateKey`'s and `fetchOne`'s call order irrelevant? 
- Is there a particular reason why the parameter of `usingCandidateKey` an object and not an array? (just out of curiosity)

**Comment by AnyhowStep** [comment]
> How difficult would it be to make usingCandidateKey's and fetchOne's call order irrelevant?

It would require extra copy-pasting/effort, and I don't see much benefit in it.

The main reason I brought up the `usingX` approach was to reduce the amount of inelegant copy-pasting I've been doing internally. I only decided against it because I was worried that it might make the API harder to use.

But it seems like at least two people think `usingX` is good for usability, too. So, win-win.

Making the call order irrelevant would undo all the "savings" I'd get from `usingX`.

-----

Another reason I'd avoid supporting it is because I want to preserve an undocumented rule I have for the semantics of this library.

A lot of this library uses the builder pattern. Normally, with a builder pattern, you have,
```ts
myBuilder
  .stepA()
  .stepB()
  .stepC()
  /*snip*/
  .build();
```

Whereas with this library, you have,
```ts
myBuilder
  .stepA()
  .stepB()
  .stepC()
  /*snip*/
  .build(connection);
```

You have a bunch a "configuration steps", and a final "build step".

In general, you can reorder calls to each "configuration step" in any way you want (with exceptions every now and then). You could have `A, B, C` or `C, B, A` or `C, A, B`. Or even `C, B, B, A, B`.

For this library, when you finally "build", I want it to actually execute the query at that point. And to minimize confusion, it's better if each phrase used for a "build" method is only used to mean "build and execute the query now".

If I made the call order irrelevant, we would have cases where a "build" method can mean a "configuration step", instead, in some cases. My personal opinion is that it just adds cognitive overhead (both to the library dev and downstream users).

-----

> Is there a particular reason why the parameter of usingCandidateKey is an object and not an array? (just out of curiosity)

Candidate keys can be made of one, two, or more columns.

-----

### Candidate Key with One Column

If you allow array input,
```ts
myTable.usingCandidateKey([myCandidateKey0])
```
If you allow the value to be used directly,
```ts
myTable.usingCandidateKey(myCandidateKey0)
```

-----

This library stresses safety as much as possible. Sometimes, safety means "safety from typos".

I want to avoid scenarios like these,
```ts
myTable.usingCandidateKey([theWrongCandidateKey])
myTable.usingCandidateKey([someObj.theWrongCandidateKey])

myTable.usingCandidateKey(theWrongCandidateKey)
myTable.usingCandidateKey(someObj.theWrongCandidateKey)
```

In this case, it is blatantly obvious that we have used the wrong candidate key. When queries start getting complex, the probability of a typo goes up, too.

Most of the time, I find that I have a `someObj`, that has the candidate key, rather than just a `myCandidateKey : number` variable. So, most of the time, we'd have this,

```ts
myTable.usingCandidateKey([someObj.theWrongCandidateKey])
myTable.usingCandidateKey(someObj.theWrongCandidateKey)
```

For example, with `express`, your `request.body/params/query` object might contain the properties `myCandidateKey0, myCandidateKey1, etc.`

-----

With the object approach, we have this,
```ts
myTable.usingCandidateKey(express.params)
```

With the array/value approach, we have this,
```ts
myTable.usingCandidateKey([express.params.myCandidateKey0])
//Whoops, I made a typo.
//I meant `myCandidateKey0`, not `otherCandidateKey0`
myTable.usingCandidateKey(express.params.otherCandidateKey0)
```

The array/value approach is more error-prone, and requires more typing (based on the assumption most code will work with objects, rather than raw values).

-----

It is possible to have a custom value you wish to use, or for property names on an object to not be the same as the column aliases of the candidate key.

For example,
```ts
user.usingCandidateKey({
  userId : row.createdByUserId,
})
```

Or,
```ts
user.usingCandidateKey([row.createdByUserId])
user.usingCandidateKey(row.createdByUserId)
```

In this case, **both are equally error-prone**. And the object approach requires more typing.
However, this use case happens less frequently than the same-name-same-alias scenario.

And I find the explicit `userId` annotation helps with readability.

-----

### Candidate Key with Two or More Columns

This is where things go very wrong with the array approach. Our number of possible typos has gone up. And we have a new problem.

Which is correct? Which is incorrect? Which column of the candidate key comes first? Does the first column mean `xxxId`? Or does it mean `yyyId`? It's hard to tell.
```ts
myTable.usingCandidateKey([someObj.myCandidateKey0, someObj.myCandidateKey1])
myTable.usingCandidateKey([someObj.myCandidateKey1, someObj.myCandidateKey0])
```

With the object approach,
```ts
myTable.usingCandidateKey(someObj)
```

-----

I could elaborate more about this but hopefully the above is convincing enough.

TL;DR

+ Arrays are more error-prone for common scenarios
+ Arrays are more verbose for common scenarios

**Comment by pachoclo** [comment]
The explanations to both questions are convincing. You clearly thought this through. 👍🏻

**Comment by AnyhowStep** [comment]
If I have,
```ts
myTable
  .where(() => /*arbitrary WHERE clause*/)
  .fetchOne(connection)
```

I'm wondering if I should use,
```ts
myTable
  .whereEqPrimaryKey({ pk0, pk1 })
  .fetchOne(connection)
```

Instead of,
```ts
myTable
  .usingPrimaryKey({ pk0, pk1 })
  .fetchOne(connection)
```

Because it's really a convenience method for,
```ts
myTable
  .where(() => eqPrimaryKey(
    myTable,
    { pk0, pk1 }
  ))
  .fetchOne(connection)
```

**Comment by AnyhowStep** [comment]
Finally fucking done

**Comment by G-Rath** [comment]
@AnyhowStep now now, language 😜
